fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBMetaDataStore.java

```diff
@@ -420,6 +429,181 @@ public void saveRecordMetaData(@Nonnull RecordMetaDataProto.MetaData metaDataPro
+     */
+    public void addIndex(@Nonnull String recordType, @Nonnull String indexName, @Nonnull String fieldName) {
+        context.asyncToSync(FDBStoreTimer.Waits.WAIT_ADD_INDEX, addIndexAsync(recordType, indexName, fieldName));
+    }
+
+    /**
+     * Add a new index to the record meta-data.
+     * @param recordType the name of the record type
+     * @param indexName the name of the new index
+     * @param indexExpression the root expression of the index
+     */
+    public void addIndex(@Nonnull String recordType, @Nonnull String indexName, @Nonnull KeyExpression indexExpression) {
+        context.asyncToSync(FDBStoreTimer.Waits.WAIT_ADD_INDEX, addIndexAsync(recordType, indexName, indexExpression));
+    }
+
+    /**
+     * Add a new index to the record meta-data.
+     * @param recordType the name of the record type
+     * @param index the new index to be added
+     */
+    public void addIndex(@Nonnull String recordType, @Nonnull Index index) {
+        context.asyncToSync(FDBStoreTimer.Waits.WAIT_ADD_INDEX, addIndexAsync(recordType, index));
+    }
+
+    /**
+     * Add a new index to the record meta-data asynchronously.
+     * @param recordType the name of the record type
+     * @param indexName the name of the new index
+     * @param fieldName the field to be indexed
+     * @return a future that completes when the index is added
+     */
+    public CompletableFuture<Void> addIndexAsync(@Nonnull String recordType, @Nonnull String indexName, @Nonnull String fieldName) {
+        return addIndexAsync(recordType, new Index(indexName, fieldName));
+    }
+
+    /**
+     * Add a new index to the record meta-data asynchronously.
+     * @param recordType the name of the record type
+     * @param indexName the name of the new index
+     * @param indexExpression the root expression of the index
+     * @return a future that completes when the index is added
+     */
+    public CompletableFuture<Void> addIndexAsync(@Nonnull String recordType, @Nonnull String indexName,
+                                                 @Nonnull KeyExpression indexExpression) {
+        return addIndexAsync(recordType, new Index(indexName, indexExpression));
+    }
+
+    /**
+     * Add a new index to the record meta-data asynchronously.
+     * @param recordType the name of the record type
+     * @param index the index to be added
+     * @return a future that completes when the index is added
+     */
+    public CompletableFuture<Void> addIndexAsync(@Nonnull String recordType, @Nonnull Index index) {
+        return loadCurrentProto().thenCompose(metaDataProto -> {
+            RecordMetaDataBuilder recordMetaDataBuilder = createMetaDataBuilder(metaDataProto);
+            recordMetaDataBuilder.addIndex(recordType, index);
+            return saveAndSetCurrent(recordMetaDataBuilder.getRecordMetaData().toProto());
+        });
+    }
+
+    /**
+     * Adds a new index to the record meta-data that contains multiple record types.
```

Comment:
Roughly half of the methods here say "Adds a new index" and the other half say "Add a new index". I don't know which one I prefer, but I think they should all be the same.